File: mcp-code-intelligence-nodejs/src/orchestration/routing/table.ts
/**
 * Routing table — O(1) tool-to-server mapping.
 * Behavioral parity with Kotlin RoutingTable.kt.
 */

export interface RouteEntry {
  serverName: string;
  isNative: boolean;
}

export class RoutingTable {
  private routes = new Map<string, RouteEntry>();

  /** Rebuild routing table from native tools and child server tools. */
  rebuild(nativeNames: Set<string>, childByServer: Map<string, string[]>): void {
    this.routes.clear();
    for (const name of nativeNames) {
      this.routes.set(name, { serverName: 'native', isNative: true });
    }
    for (const [source, tools] of childByServer) {
      const serverName = source.replace(/^child:/, '');
      for (const toolName of tools) {
        if (!this.routes.has(toolName)) {
          this.routes.set(toolName, { serverName, isNative: false });
        }
      }
    }
  }

  /** Resolve tool name to route entry. Returns null if not found. */
  resolve(toolName: string): RouteEntry | null {
    return this.routes.get(toolName) ?? null;
  }

  /** Add a single route entry (for dynamically discovered nested tools). */
  addRoute(toolName: string, serverName: string): void {
    if (!this.routes.has(toolName)) {
      this.routes.set(toolName, { serverName, isNative: false });
    }
  }
}

File: backend/mcp-code-intelligence-python/src/mcp_code_intel/orchestration/routing/table.py
"""Routing table — O(1) tool-to-server mapping.

Behavioral parity with Kotlin RoutingTable.kt.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass
class RouteEntry:
    """Routing entry for a single tool."""

    server_name: str
    is_native: bool = False


class RoutingTable:
    """O(1) lookup from tool name to server name."""

    def __init__(self) -> None:
        self._routes: dict[str, RouteEntry] = {}

    def rebuild(self, native_names: set[str], child_by_server: dict[str, list[str]]) -> None:
        """Rebuild routing table from native tools and child server tools."""
        self._routes.clear()
        for name in native_names:
            self._routes[name] = RouteEntry(server_name="native", is_native=True)
        for source, tools in child_by_server.items():
            server_name = source.removeprefix("child:")
            for tool_name in tools:
                if tool_name not in self._routes:
                    self._routes[tool_name] = RouteEntry(server_name=server_name)

    def resolve(self, tool_name: str) -> RouteEntry | None:
        """Resolve tool name to route entry. Returns None if not found."""
        return self._routes.get(tool_name)

    def add_route(self, tool_name: str, server_name: str) -> None:
        """Add a single route entry (for dynamically discovered tools)."""
        if tool_name not in self._routes:
            self._routes[tool_name] = RouteEntry(server_name=server_name)
